src/main/kotlin/io/emeraldpay/dshackle/rpc/SubscribeNodeStatus.kt

```diff
@@ -0,0 +1,139 @@
+package io.emeraldpay.dshackle.rpc
+
+import io.emeraldpay.api.proto.BlockchainOuterClass
+import io.emeraldpay.api.proto.BlockchainOuterClass.NodeDescription
+import io.emeraldpay.api.proto.BlockchainOuterClass.NodeStatus
+import io.emeraldpay.api.proto.BlockchainOuterClass.NodeStatusResponse
+import io.emeraldpay.api.proto.BlockchainOuterClass.SubscribeNodeStatusRequest
+import io.emeraldpay.api.proto.Common
+import io.emeraldpay.dshackle.Chain
+import io.emeraldpay.dshackle.upstream.CurrentMultistreamHolder
+import io.emeraldpay.dshackle.upstream.Upstream
+import io.emeraldpay.dshackle.upstream.UpstreamAvailability
+import org.springframework.stereotype.Service
+import reactor.core.publisher.Flux
+import reactor.core.publisher.Mono
+import reactor.core.publisher.SignalType
+import reactor.core.publisher.Sinks
+import java.time.Duration
+import java.util.concurrent.ConcurrentHashMap
+
+@Service
+class SubscribeNodeStatus(
+    private val multistreams: CurrentMultistreamHolder
+) {
+
+    companion object {
+        private val RETRY_TIMEOUT = Duration.ofSeconds(10)
+    }
+
+    fun subscribe(req: Mono<SubscribeNodeStatusRequest>): Flux<NodeStatusResponse> =
+        req.flatMapMany { request ->
+            val knownUpstreams = ConcurrentHashMap<String, Boolean>()
```

Comment:
What happens if some upstream becomes unavailable and then returned back? Looks like it can be ignored by that map

Reply:
now it's OK